Web Inspector: UAF needs to be prevented for RemoteConnectionToTarget::m_target member variable

robert-jenner opened the pull request:
#### 2a508351f154cb9a1ff2789d0fceb615f4dd2d7d
<pre>
Web Inspector: UAF needs to be prevented for RemoteConnectionToTarget::m_target member variable
<a href="https://bugs.webkit.org/show_bug.cgi?id=276192">https://bugs.webkit.org/show_bug.cgi?id=276192</a>
<a href="https://rdar.apple.com/129782183">rdar://129782183</a>

Reviewed by Geoffrey Garen.

`RemoteConnectionToTarget::m_target` was a raw pointer and used from multiple thread.
We have evidence from the radar that `m_target` could be used-after-free in
`RemoteConnectionToTarget::close()`.

To address the issue, I am updating `m_target` to use a ThreadSafeWeakPtr instead
of a raw pointer so that it gets nulled out on target destruction, and still safe
to use from multiple thread.

Of course, using ThreadSafeWeakPtr required updating `RemoteControllableTarget` to
subclass `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr` and be refcounted. This
required a decent amount of refactoring.

I am also adding WTF_GUARDED_BY_LOCK for the m_target so that the compiler tells
us when we&apos;re using it without a lock. It did find a few problems which I fixed.

* Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
(Inspector::RemoteConnectionToTarget::setup):
(Inspector::RemoteConnectionToTarget::sendMessageToTarget): Deleted.
(Inspector::RemoteConnectionToTarget::close): Deleted.
(Inspector::RemoteConnectionToTarget::targetClosed): Deleted.
(Inspector::RemoteConnectionToTarget::targetIdentifier const): Deleted.
(Inspector::RemoteConnectionToTarget::sendMessageToFrontend): Deleted.
* Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.h:
* Source/JavaScriptCore/inspector/remote/RemoteControllableTarget.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteConnectionToTargetCocoa.mm:
(Inspector::RemoteConnectionToTarget::targetIdentifier const):
(Inspector::RemoteConnectionToTarget::setup):
(Inspector::RemoteConnectionToTarget::close):
(Inspector::RemoteConnectionToTarget::sendMessageToTarget):
(Inspector::RemoteConnectionToTarget::sendMessageToFrontend):
(Inspector::RemoteConnectionToTarget::setupRunLoop):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::~JSGlobalObject):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::inspectorDebuggable):
* Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.cpp:
(JSC::JSGlobalObjectDebuggable::create):
(JSC::JSGlobalObjectDebuggable::JSGlobalObjectDebuggable):
(JSC::JSGlobalObjectDebuggable::name const):
(JSC::JSGlobalObjectDebuggable::connect):
(JSC::JSGlobalObjectDebuggable::disconnect):
(JSC::JSGlobalObjectDebuggable::dispatchMessageFromRemote):
(JSC::JSGlobalObjectDebuggable::pauseWaitingForAutomaticInspection):
(JSC::JSGlobalObjectDebuggable::globalObjectDestroyed):
* Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
(WebCore::Page::~Page):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageDebuggable.cpp:
(WebCore::PageDebuggable::create):
(WebCore::PageDebuggable::PageDebuggable):
(WebCore::PageDebuggable::name const):
(WebCore::PageDebuggable::url const):
(WebCore::PageDebuggable::hasLocalDebugger const):
(WebCore::PageDebuggable::connect):
(WebCore::PageDebuggable::disconnect):
(WebCore::PageDebuggable::dispatchMessageFromRemote):
(WebCore::PageDebuggable::setIndicating):
(WebCore::PageDebuggable::detachFromPage):
* Source/WebCore/page/PageDebuggable.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::create):
(WebCore::ServiceWorkerDebuggable::connect):
(WebCore::ServiceWorkerDebuggable::disconnect):
(WebCore::ServiceWorkerDebuggable::dispatchMessageFromRemote):
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::Debuggable::create):
(WebKit::WebAutomationSession::Debuggable::Debuggable):
(WebKit::WebAutomationSession::Debuggable::sessionDestroyed):
(WebKit::WebAutomationSession::Debuggable::name const):
(WebKit::WebAutomationSession::Debuggable::dispatchMessageFromRemote):
(WebKit::WebAutomationSession::Debuggable::connect):
(WebKit::WebAutomationSession::Debuggable::disconnect):
(WebKit::WebAutomationSession::WebAutomationSession):
(WebKit::WebAutomationSession::~WebAutomationSession):
(WebKit::WebAutomationSession::connect):
(WebKit::WebAutomationSession::init):
(WebKit::WebAutomationSession::isPaired const):
(WebKit::WebAutomationSession::isPendingTermination const):
(WebKit::WebAutomationSession::terminate):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Inspector/WebPageDebuggable.cpp:
(WebKit::WebPageDebuggable::create):
(WebKit::WebPageDebuggable::WebPageDebuggable):
(WebKit::WebPageDebuggable::detachFromPage):
(WebKit::WebPageDebuggable::name const):
(WebKit::WebPageDebuggable::url const):
(WebKit::WebPageDebuggable::hasLocalDebugger const):
(WebKit::WebPageDebuggable::connect):
(WebKit::WebPageDebuggable::disconnect):
(WebKit::WebPageDebuggable::dispatchMessageFromRemote):
(WebKit::WebPageDebuggable::setIndicating):
* Source/WebKit/UIProcess/Inspector/WebPageDebuggable.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::~WebPageProxy):
(WebKit::WebPageProxy::close):
* Source/WebKit/UIProcess/WebPageProxy.h:

Originally-landed-as: 280938.63@safari-7619-branch (996e25f1fd95). <a href="https://rdar.apple.com/136111011">rdar://136111011</a>
Canonical link: <a href="https://commits.webkit.org/285039@main">https://commits.webkit.org/285039@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fdf5dba5871c3216f76adc6e5359b64dae56fd0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71319 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24092 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75427 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22524 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73434 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58532 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22343 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56368 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14832 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46079 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61457 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36806 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42748 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18917 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20865 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64444 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/64645 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19283 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77149 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70568 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15553 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18476 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15596 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61492 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12211 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5843 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92354 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10935 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46532 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1311 "Found 9 new failures in page/PageDebuggable.h, UIProcess/Automation/WebAutomationSession.cpp, workers/service/context/ServiceWorkerInspectorProxy.h, UIProcess/Automation/WebAutomationSession.h and found 1 fixed file: workers/service/context/ServiceWorkerDebuggable.h") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20371 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47603 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48886 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47345 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/Inspector/WebPageDebuggable.cpp:
```diff
@@ -31,67 +31,97 @@
 
 WebPageDebuggable::WebPageDebuggable(WebPageProxy& page)
-    : m_page(page)
+    : protectedPage()
```
Please revert this line change.

robert-jenner replied:
Resolved.

cdumez commented on Source/WebKit/UIProcess/Inspector/WebPageDebuggable.cpp:
```diff
@@ -31,67 +31,97 @@
 {
-    if (!m_page->mainFrame())
-        return String();
+    protectedPage() = nullptr;
```
This should be `m_page = nullptr;`

robert-jenner replied:
Resolved.

cdumez commented on Source/WebKit/UIProcess/Inspector/WebPageDebuggable.h:
```diff
@@ -57,7 +57,11 @@ class WebPageDebuggable final : public Inspector::RemoteInspectionTarget {
+
     Ref<WebPageProxy> protectedPage() const;
 
     WeakRef<WebPageProxy> m_page;
```
This should be:
```
WeakPtr<WebPageProxy> m_page;
```

cdumez commented on Source/WebKit/UIProcess/Inspector/WebPageDebuggable.h:
```diff
@@ -57,7 +57,11 @@ class WebPageDebuggable final : public Inspector::RemoteInspectionTarget {
+    void detachFromPage();
+
 private:
+    WeakPtr<WebPageProxy> m_page;
```
Missing `explicit WebPageDebuggable(WebPageProxy&);` constructor.

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
```diff
@@ -44,21 +44,27 @@ RemoteConnectionToTarget::~RemoteConnectionToTarget() = default;
+        Locker locker { m_targetMutex };
         m_target = nullptr;
-    } else if (auto* target = dynamicDowncast<RemoteInspectionTarget>(m_target)) {
+    } else if (auto* target = dynamicDowncast<RemoteInspectionTarget>(*target) {
```
`} else if (auto* inspectionTarget = dynamicDowncast<RemoteInspectionTarget>(*target) {`

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
```diff
@@ -44,21 +44,27 @@ RemoteConnectionToTarget::~RemoteConnectionToTarget() = default;
         m_target = nullptr;
-    } else if (auto* target = dynamicDowncast<RemoteInspectionTarget>(m_target)) {
+    } else if (auto* target = dynamicDowncast<RemoteInspectionTarget>(*target) {
         target->connect(*this, isAutomaticInspection, automaticallyPause);
```
`inspectionTarget->connect(*this, isAutomaticInspection, automaticallyPause);`

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
```diff
@@ -44,21 +44,27 @@ RemoteConnectionToTarget::~RemoteConnectionToTarget() = default;
 
         RemoteInspector::singleton().updateTargetListing(targetIdentifier);
-    } else if (auto* target = dynamicDowncast<RemoteAutomationTarget>(m_target)) {
+    } else if (auto* target = dynamicDowncast<RemoteAutomationTarget>(*target)) {
```
`} else if (auto* inspectionTarget = dynamicDowncast<RemoteAutomationTarget>(*target)) {`

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
```diff
@@ -44,21 +44,27 @@ RemoteConnectionToTarget::~RemoteConnectionToTarget() = default;
         RemoteInspector::singleton().updateTargetListing(targetIdentifier);
-    } else if (auto* target = dynamicDowncast<RemoteAutomationTarget>(m_target)) {
+    } else if (auto* target = dynamicDowncast<RemoteAutomationTarget>(*target)) {
         target->connect(*this);
```
`inspectionTarget->connect(*this);`

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
```diff
@@ -107,15 +112,21 @@ void RemoteConnectionToTarget::targetClosed()
+    std::optional<TargetID> targetIdentifier;
+    {
+        Locker locker { m_targetMutex };
+        RefPtr target = m_target.get()
```
Missing semicolon at the end.

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
```diff
@@ -107,15 +112,21 @@ void RemoteConnectionToTarget::targetClosed()
+            return;
+        targetIdentifier = target->targetIdentifier();
+    }
+    RemoteInspector::singleton().sendMessageToRemote(targetIdentifier, message);
```
`RemoteInspector::singleton().sendMessageToRemote(*targetIdentifier, message);`

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
```diff
@@ -44,22 +44,28 @@ RemoteConnectionToTarget::~RemoteConnectionToTarget() = default;
         m_target = nullptr;
-    } else if (auto* target = dynamicDowncast<RemoteInspectionTarget>(m_target)) {
-        target->connect(*this, isAutomaticInspection, automaticallyPause);
+    } else if (auto* inspectionTarget = dynamicDowncast<RemoteInspectionTarget>(*target) {
```
`} else if (auto* inspectionTarget = dynamicDowncast<RemoteInspectionTarget>(*target)) {`

There is a missing parenthesis.

cdumez commented on Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.cpp:
```diff
@@ -35,49 +35,98 @@
 void JSGlobalObjectDebuggable::dispatchMessageFromRemote(String&& message)
 {
-    JSLockHolder locker(&m_globalObject.vm());
+    callOnGlobalObjectRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()] {
```
`callOnGlobalObjectRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()]() mutable {`

dcrousso commented on Source/WebCore/page/PageDebuggable.cpp:
```diff
@@ -34,67 +34,104 @@
-    if (!localMainFrame)
-        return String();
+    String name;
+    callOnMainThreadAndWait([this, protectedThis = Ref { *this }, &name] {
```
can you explain a bit of why this is necessary too?  i can understand making `m_target` guarded by a `Lock` and even the changes to `JSGlobalObjectDebuggable` just in case someone access a `JSContext` from a thread other than where it was created, but this shouldn't be reachable by anyone outside of WebKit and should only ever be created on the main thread (since it lives in the WebProcess).  what's going on for this?

cdumez replied:
This is a cherry-pick from the security branch. I wrote the fix but do not remember exactly all the details. I'm not a WebInspector expert so if you feel like some changes in the patch are not needed, I encourage you to follow-up (likely adding assertions to guarantee this is not called off the main thread).

What I remember is that we had crashes because some of the functions was called off the main thread and the code wasn't thread safe. That said, it is entirely possible I updated code that didn't technically need to be by trying to fix surrounding code.

dcrousso replied:
can you provide any more info like repro steps or backtraces?  my understanding of how all this works is that this isn't necessary, but i also could be wrong and would love to learn if that's the case :)

cdumez replied:
We don't have reproduction steps and the crash in the wild looked like:
```
3 __cxa_pure_virtual
                  5 Inspector::RemoteConnectionToTarget::close()::$_0::operator()() const
            ==> 5 WTF::Detail::CallableWrapper<Inspector::RemoteConnectionToTarget::close()::$_0, void>::call() <==
                      5 WTF::Function<void ()>::operator()() const
                        5 Inspector::RemoteTargetHandleRunSourceGlobal(void*)
                          5 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__
```

dcrousso replied:
hmm do you know if this was from a `WKWebView` or a `JSContext`?

cdumez replied:
It was crashing in the SafariViewService process, it's all I know. As far as I know, the crash did go away with this fix.

cdumez commented on Source/WebCore/page/PageDebuggable.cpp:
```diff
@@ -34,67 +34,104 @@
 void PageDebuggable::dispatchMessageFromRemote(String&& message)
 {
-    m_page.inspectorController().dispatchMessageFromFrontend(WTFMove(message));
+    callOnMainThreadAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()] {
```
`callOnMainThreadAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()]() mutable {`

cdumez commented on Source/WebKit/UIProcess/Inspector/WebPageDebuggable.cpp:
```diff
@@ -39,59 +40,93 @@ using namespace Inspector;
 void WebPageDebuggable::dispatchMessageFromRemote(String&& message)
 {
-    protectedPage()->inspectorController().dispatchMessageFromFrontend(WTFMove(message));
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()] {
```
`callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()]() mutable {`

cdumez commented on Source/WebKit/UIProcess/Inspector/WebPageDebuggable.h:
```diff
@@ -57,10 +57,12 @@ class WebPageDebuggable final : public Inspector::RemoteInspectionTarget {
+    explicit WebPageDebuggable(WebPageProxy&);
 
-    WeakRef<WebPageProxy> m_page;
+    WeakPtr<WebPageProxy> m_page;
```
We should replace the `#include <wtf/WeakRef.h>` with `#include <wtf/WeakPtr.h>` at the top of this file.

cdumez commented on Source/JavaScriptCore/inspector/remote/RemoteControllableTarget.h:
```diff
@@ -41,7 +42,7 @@ class FrontendChannel;
-class RemoteControllableTarget {
+class JS_EXPORT_PRIVATE RemoteControllableTarget : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<RemoteControllableTarget> {
 public:
     JS_EXPORT_PRIVATE RemoteControllableTarget();
```
We need to remove `JS_EXPORT_PRIVATE ` here and on the 3 lines below, for the Windows build.

robert-jenner committed:
Web Inspector: UAF needs to be prevented for RemoteConnectionToTarget::m_target member variable
https://bugs.webkit.org/show_bug.cgi?id=276192
rdar://129782183

Reviewed by Geoffrey Garen.

`RemoteConnectionToTarget::m_target` was a raw pointer and used from multiple thread.
We have evidence from the radar that `m_target` could be used-after-free in
`RemoteConnectionToTarget::close()`.

To address the issue, I am updating `m_target` to use a ThreadSafeWeakPtr instead
of a raw pointer so that it gets nulled out on target destruction, and still safe
to use from multiple thread.

Of course, using ThreadSafeWeakPtr required updating `RemoteControllableTarget` to
subclass `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr` and be refcounted. This
required a decent amount of refactoring.

I am also adding WTF_GUARDED_BY_LOCK for the m_target so that the compiler tells
us when we're using it without a lock. It did find a few problems which I fixed.

* Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
(Inspector::RemoteConnectionToTarget::setup):
(Inspector::RemoteConnectionToTarget::sendMessageToTarget): Deleted.
(Inspector::RemoteConnectionToTarget::close): Deleted.
(Inspector::RemoteConnectionToTarget::targetClosed): Deleted.
(Inspector::RemoteConnectionToTarget::targetIdentifier const): Deleted.
(Inspector::RemoteConnectionToTarget::sendMessageToFrontend): Deleted.
* Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.h:
* Source/JavaScriptCore/inspector/remote/RemoteControllableTarget.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteConnectionToTargetCocoa.mm:
(Inspector::RemoteConnectionToTarget::targetIdentifier const):
(Inspector::RemoteConnectionToTarget::setup):
(Inspector::RemoteConnectionToTarget::close):
(Inspector::RemoteConnectionToTarget::sendMessageToTarget):
(Inspector::RemoteConnectionToTarget::sendMessageToFrontend):
(Inspector::RemoteConnectionToTarget::setupRunLoop):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::~JSGlobalObject):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::inspectorDebuggable):
* Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.cpp:
(JSC::JSGlobalObjectDebuggable::create):
(JSC::JSGlobalObjectDebuggable::JSGlobalObjectDebuggable):
(JSC::JSGlobalObjectDebuggable::name const):
(JSC::JSGlobalObjectDebuggable::connect):
(JSC::JSGlobalObjectDebuggable::disconnect):
(JSC::JSGlobalObjectDebuggable::dispatchMessageFromRemote):
(JSC::JSGlobalObjectDebuggable::pauseWaitingForAutomaticInspection):
(JSC::JSGlobalObjectDebuggable::globalObjectDestroyed):
* Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
(WebCore::Page::~Page):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageDebuggable.cpp:
(WebCore::PageDebuggable::create):
(WebCore::PageDebuggable::PageDebuggable):
(WebCore::PageDebuggable::name const):
(WebCore::PageDebuggable::url const):
(WebCore::PageDebuggable::hasLocalDebugger const):
(WebCore::PageDebuggable::connect):
(WebCore::PageDebuggable::disconnect):
(WebCore::PageDebuggable::dispatchMessageFromRemote):
(WebCore::PageDebuggable::setIndicating):
(WebCore::PageDebuggable::detachFromPage):
* Source/WebCore/page/PageDebuggable.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::create):
(WebCore::ServiceWorkerDebuggable::connect):
(WebCore::ServiceWorkerDebuggable::disconnect):
(WebCore::ServiceWorkerDebuggable::dispatchMessageFromRemote):
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::Debuggable::create):
(WebKit::WebAutomationSession::Debuggable::Debuggable):
(WebKit::WebAutomationSession::Debuggable::sessionDestroyed):
(WebKit::WebAutomationSession::Debuggable::name const):
(WebKit::WebAutomationSession::Debuggable::dispatchMessageFromRemote):
(WebKit::WebAutomationSession::Debuggable::connect):
(WebKit::WebAutomationSession::Debuggable::disconnect):
(WebKit::WebAutomationSession::WebAutomationSession):
(WebKit::WebAutomationSession::~WebAutomationSession):
(WebKit::WebAutomationSession::connect):
(WebKit::WebAutomationSession::init):
(WebKit::WebAutomationSession::isPaired const):
(WebKit::WebAutomationSession::isPendingTermination const):
(WebKit::WebAutomationSession::terminate):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Inspector/WebPageDebuggable.cpp:
(WebKit::WebPageDebuggable::create):
(WebKit::WebPageDebuggable::WebPageDebuggable):
(WebKit::WebPageDebuggable::detachFromPage):
(WebKit::WebPageDebuggable::name const):
(WebKit::WebPageDebuggable::url const):
(WebKit::WebPageDebuggable::hasLocalDebugger const):
(WebKit::WebPageDebuggable::connect):
(WebKit::WebPageDebuggable::disconnect):
(WebKit::WebPageDebuggable::dispatchMessageFromRemote):
(WebKit::WebPageDebuggable::setIndicating):
* Source/WebKit/UIProcess/Inspector/WebPageDebuggable.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::~WebPageProxy):
(WebKit::WebPageProxy::close):
* Source/WebKit/UIProcess/WebPageProxy.h:

Originally-landed-as: 280938.63@safari-7619-branch (996e25f1fd95). rdar://136111011
Canonical link: https://commits.webkit.org/285039@main
